File: eHour-persistence/src/main/java/net/rrm/ehour/persistence/user/dao/UserDaoHibernateImpl.java
/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

package net.rrm.ehour.persistence.user.dao;

import com.google.common.base.Optional;
import net.rrm.ehour.domain.User;
import net.rrm.ehour.domain.UserDepartment;
import net.rrm.ehour.persistence.dao.AbstractGenericDaoHibernateImpl;
import org.hibernate.Query;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository("userDao")
public class UserDaoHibernateImpl extends AbstractGenericDaoHibernateImpl<User, Integer> implements UserDao {
    private static final Optional<String> CACHEREGION = Optional.of("query.User");

    public UserDaoHibernateImpl() {
        super(User.class);
    }

    @Override
    public User findByUsername(String username) {
        List<User> l = findByNamedQueryAndNamedParam("User.findByUsername", "username", username, CACHEREGION);

        return l.size() > 0 ? l.get(0) : null;
    }

    @Override
    @SuppressWarnings("unchecked")
    public List<User> findUsers(boolean onlyActive) {
        return onlyActive ? findByNamedQuery("User.findActiveUsers", CACHEREGION) : findAll();
    }

    @SuppressWarnings("unchecked")
    @Override
    public List<User> findActiveUsers() {
        return findUsers(true);
    }

    @Override
    public List<User> findUsersForDepartments(List<UserDepartment> departments, boolean onlyActive) {
        String hql;
        String[] paramKeys;
        Object[] paramValues;

        hql = (onlyActive) ? "User.findActiveForDepartment" :
                "User.findForDepartment";

        paramKeys = new String[]{"departments"};
        paramValues = new Object[]{departments.toArray()};

        return findByNamedQueryAndNamedParam(hql, paramKeys, paramValues, CACHEREGION);
    }

    @SuppressWarnings("unchecked")
    @Override
    public List<User> findAllActiveUsersWithEmailSet() {
        return findByNamedQuery("User.findAllActiveUsersWithEmailSet", CACHEREGION);
    }

    @Override
    public void deletePmWithoutProject() {
        Query q = getSession().getNamedQuery("User.deletePMsWithoutProject");
        q.executeUpdate();
    }
}

File: eHour-persistence/src/main/scala/net/rrm/ehour/persistence/user/dao/UserDaoHibernateImpl.scala
package net.rrm.ehour.persistence.user.dao

import java.util

import com.google.common.collect.Lists
import net.rrm.ehour.data.LegacyUserDepartment
import net.rrm.ehour.domain.User
import net.rrm.ehour.persistence.dao.AbstractGenericDaoHibernateImpl
import net.rrm.ehour.persistence.retry.ExponentialBackoffRetryPolicy
import org.hibernate.exception.SQLGrammarException
import org.hibernate.transform.Transformers
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

@Repository("userDao")
class UserDaoHibernateImpl extends AbstractGenericDaoHibernateImpl[Integer, User](classOf[User]) with UserDao {
  private final val CacheRegion = Some("query.User")

  @Transactional(readOnly = true)
  override def findByUsername(username: String): User = {
    val results = ExponentialBackoffRetryPolicy retry findByNamedQuery("User.findByUsername", "username", username, CacheRegion)

    if (results.size > 0) results.get(0) else null
  }

  @Transactional(readOnly = true)
  override def findUsers(onlyActive: Boolean): util.List[User] = {
    if (onlyActive)
      findByNamedQuery("User.findActiveUsers", CacheRegion)
    else
      findAll()
  }

  @Transactional(readOnly = true)
  override def findActiveUsers(): util.List[User] = findUsers(onlyActive = true)

  @Transactional(readOnly = true)
  override def findAllActiveUsersWithEmailSet(): util.List[User] = findByNamedQuery("User.findAllActiveUsersWithEmailSet", CacheRegion)

  @Transactional
  override def deletePmWithoutProject() {
    ExponentialBackoffRetryPolicy retry getSession.getNamedQuery("User.deletePMsWithoutProject").executeUpdate
  }

  @Transactional
  override def findLegacyUserDepartments(): util.List[LegacyUserDepartment] = {
    try {
      getSession.createSQLQuery("SELECT USER_ID as userId, DEPARTMENT_ID as departmentId FROM USERS WHERE DEPARTMENT_ID IS NOT NULL")
        .addScalar("userId")
        .addScalar("departmentId")
        .setResultTransformer(Transformers.aliasToBean(classOf[LegacyUserDepartment]))
        .list()
        .asInstanceOf[util.List[LegacyUserDepartment]]
    } catch {
      case grammar: SQLGrammarException => Lists.newArrayList()
    }
  }
}